tt_metal/distributed/mesh_program.hpp

```diff
@@ -0,0 +1,134 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "mesh_device.hpp"
+#include "tt_metal/impl/program/program.hpp"
+
+namespace tt::tt_metal {
+// Mesh buffer, Mesh Op, Compilation
+class MeshProgram {
+   public:
+        MeshProgram(std::size_t num_devices);
+        ~MeshProgram() = default;
+        Program& at(std::size_t device_index);
+        std::vector<uint32_t> get_sem_base_addr(std::shared_ptr<MeshDevice> mesh_device, CoreCoord logical_core, CoreType core_type) const;
+
+        template<typename T>
+        T distributed_impl_(const std::function<T(Program&)>& callable) {
+            if constexpr (std::is_same<T, void>::value) {
+                for (std::size_t program_idx = 0; program_idx < this->programs.size(); program_idx++) {
+                    callable(*this->programs.at(program_idx));
+                }
+            } else {
+                for (std::size_t program_idx = 0; program_idx < this->programs.size() - 1; program_idx++) {
+                    callable(*this->programs.at(program_idx));
+                }
+                return callable(*this->programs.at(this->programs.size() -1));
+            }
+        }
+
+        template<typename T>
+        std::vector<T> distributed_impl_(const std::variant<std::function<T(Program&)>, std::function<T(Program&, Device*)>>& callable, std::shared_ptr<MeshDevice> mesh_device = nullptr) const {
```

Comment:
> Includes APIs to set MeshProgram configuration across entire MeshDevice or per device in the Mesh. APIs are analogous to Program config APIs
> Basic getter APIs to return individual programs and state across MeshProgram
> Relies on distribute_impl_ and distribute_to_mesh_device_ functions
> distribute_impl_ serves as the MeshProgram entry point on the Controller or Executor to process attributes of this data structure on host(s)
> This function is currently implemented as a simple loop, but it can be swapped out for a set of RPC calls on the Controller and asynchronous calls on the executor
> distribute_to_mesh_device_impl_ serves as the interface between the MeshProgram on host and the MeshDevice. 

The header interface should remain the same as we swap out the changes in the impl